FileGetContentsLoader.php

```diff
@@ -119,44 +134,40 @@ function ($code, $severity, $msg, $msgCode, $bytesTx, $bytesMax) use (
     }
 
     /**
-     * Parse HTTP Link headers
+     * Attempts to retrieve any Link header being offered for application/ld+json content negotiation.
      *
-     * @param array $values  An array of HTTP Link header values
+     * @param  array  $headers  An array of HTTP Link headers
      * @param  IRI  $baseIri The document's URL (used to expand relative URLs to absolutes)
-     *
-     * @return array An array of parsed HTTP Link headers
+     * 
+     * @return array  $links  A structured array of Link header data
      */
-    private function parseContextLinkHeaders(array $values, IRI $baseIri)
+    public function parseLinkHeaders(array $headers, IRI $baseIri)
     {
-        // Separate multiple links contained in a single header value
-        for ($i = 0, $total = count($values); $i < $total; $i++) {
-            if (strpos($values[$i], ',') !== false) {
-                foreach (preg_split('/,(?=([^"]*"[^"]*")*[^"]*$)/', $values[$i]) as $v) {
-                    $values[] = trim($v);
-                }
-                unset($values[$i]);
-            }
-        }
+        $links = array();
 
-        $contexts = $matches = array();
-        $trimWhitespaceCallback = function ($str) {
-            return trim($str, "\"'  \n\t");
-        };
-
-        // Split the header in key-value pairs
-        foreach ($values as $val) {
-            $part = array();
-            foreach (preg_split('/;(?=([^"]*"[^"]*")*[^"]*$)/', $val) as $kvp) {
-                preg_match_all('/<[^>]+>|[^=]+/', $kvp, $matches);
-                $pieces = array_map($trimWhitespaceCallback, $matches[0]);
-                $part[$pieces[0]] = isset($pieces[1]) ? $pieces[1] : '';
-            }
+        foreach ($headers as $header) { // Foreach individual Link header
+            foreach (explode(',', $header) as $value) { // Handle case of multiple links within a single Link header
+                if (preg_match("/<(.[^>]+)>;/", $value, $uri)) {
+                    $iri = new IRI(trim($uri[1]));
 
-            if (in_array('http://www.w3.org/ns/json-ld#context', explode(' ', $part['rel']))) {
-                $contexts[] = (string) $baseIri->resolve(trim(key($part), '<> '));
+                    $link = array('uri' => $iri->isAbsolute() ? (string) $iri : (string) $baseIri->resolve($iri));
+
+                    preg_match_all("/;\s?([A-z][^,=]+)=\"?(.[^\";]+)/", $value, $parameters);
+
+                    if (count($parameters) == 3) {
+                        $keys = $parameters[1];
+                        $values = $parameters[2];
+
+                        for ($i=0; $i < count($keys); $i++) {
+                            $link[trim($keys[$i])] = trim($values[$i]);
+                        }
+                    }
+
+                    $links[] = $link;
+                }
             }
         }
 
-        return array_values(array_unique($contexts));
+        return $links;
     }
 }
```

Comment:
Hmm... do you really need to change all of this. AFAICT, the only change necessary should be to change how the URI is handled in `$part` below and return all parts

```php
        // Split the header in key-value pairs
        foreach ($values as $val) {
            $part = array();
            foreach (preg_split('/;(?=([^"]*"[^"]*")*[^"]*$)/', $val) as $kvp) {
                preg_match_all('/<[^>]+>|[^=]+/', $kvp, $matches);
                $pieces = array_map($trimWhitespaceCallback, $matches[0]);
                $part[$pieces[0]] = isset($pieces[1]) ? $pieces[1] : '';
            }

            if (in_array('http://www.w3.org/ns/json-ld#context', explode(' ', $part['rel']))) {
                $contexts[] = (string) $baseIri->resolve(trim(key($part), '<> '));
            }
        }

        return array_values(array_unique($contexts));
```

Something like this

```php
        $result = array();

        // Split the header in key-value pairs
        foreach ($values as $val) {
            $part = array();
            foreach (preg_split('/;(?=([^"]*"[^"]*")*[^"]*$)/', $val) as $kvp) {
                preg_match_all('/<[^>]+>|[^=]+/', $kvp, $matches);
                $pieces = array_map($trimWhitespaceCallback, $matches[0]);
->              if (!isset($pieces[1])) {
->                 $part['uri'] = (string) $baseIri->resolve(trim($pieces[0], '<> '));
->              } else {
->                  $part[$pieces[0]] = $pieces[1];
->              }
            }
->          $result[] = $part;
        }

->      return $result;
```

Reply:
I've adopted this approach in latest set of commits (mostly as you suggested, but tweaked a little as my unit tests revealed a few issues 👍)